lib/mix/lib/mix/dep/adder.ex

```diff
@@ -0,0 +1,132 @@
+defmodule Mix.Dep.Adder do
+  def add(app, opts, mix_exs) do
+    dep_opts = get_version_str(opts, app) <> get_only_string(opts) <> get_runtime_string(opts)
+    new_line = "{:#{app}, #{dep_opts}}"
+
+    formatted = format_to_string(mix_exs)
+
+    if mix_exs == formatted do
+      lines = String.split(mix_exs, "\n")
+      check_for_app!(lines, app)
+
+      case add_dep_line(lines, [], new_line) do
+        new_lines when is_list(new_lines) ->
+          {:ok, new_line, new_lines |> Enum.join("\n") |> format_to_string()}
+
+        :deps_not_found ->
+          {:error, new_line, :deps_not_found}
+      end
+    else
+      {:error, new_line, :not_formatted}
+    end
+  end
+
+  defp format_to_string(str),
+    do: str |> Code.format_string!() |> IO.iodata_to_binary() |> Kernel.<>("\n")
+
+  defp get_runtime_string(opts) do
+    if Keyword.has_key?(opts, :runtime) and !opts[:runtime] do
+      ", runtime: false"
+    else
+      ""
+    end
+  end
+
+  defp get_only_string(opts) do
+    case opts[:only] do
+      [] -> ""
+      [env] -> ", only: :#{env}"
+      envs -> ", only: [" <> (envs |> Enum.map(&(":" <> &1)) |> Enum.join(", ")) <> "]"
+    end
+  end
+
+  defp get_version_str(opts, app) do
+    cond do
+      path = opts[:path] ->
+        ~s|path: "#{path}"|
+
+      github = opts[:github] ->
+        ~s|github: "#{github}"| <> ref_or_tag(opts)
+
+      version = opts[:version] || latest_hex_version(app) ->
+        case version do
+          "~> " <> _ = version -> ~s|"#{version}"|
+          ">= " <> _ = version -> ~s|"#{version}"|
+          "0.0.0" -> ~s|">= 0.0.0"|
+          version -> ~s|"~> #{version}"|
+        end
+    end
+  end
+
+  defp ref_or_tag(opts) do
+    cond do
+      ref = opts[:ref] -> ~s|, ref: "#{ref}"|
+      tag = opts[:tag] -> ~s|, tag: "#{tag}"|
+      true -> ""
+    end
+  end
+
+  defp check_for_app!(lines, app) do
+    lines
+    |> Enum.find(fn line ->
+      line
+      |> String.trim()
+      |> String.starts_with?("{:" <> app)
+    end)
+    |> case do
+      nil ->
+        :ok
+
+      found ->
+        Mix.raise("#{app} already exists in mix.exs as #{String.trim(found)}")
+    end
+  end
```

Comment:
`String`-based checks are unsafe. Instead it should be done on parsing `AST`. Simplest example which would not `raise`:
```elixir
def project do
  [
    # …
    deps: [{:my_dep, …}],
    # …
  ]
end
```

Reply:
This would result in the following output in the current PR:
```
Could not find `defp deps do` in your mix.exs. Please add this line to mix.exs manually:
{:foo, ">= 0.0.0"}
```